osmnx/simplification.py

```diff
@@ -558,14 +567,25 @@ def _merge_nodes_geometric(G: nx.MultiDiGraph, tolerance: float) -> gpd.GeoSerie
     tolerance
         Buffer nodes to this distance (in graph's geometry's units) then merge
         overlapping polygons into a single polygon via unary union operation.
+        Can be a float value or a dictionary mapping node IDs to individual
+        tolerance values.
 
     Returns
     -------
     merged
         The merged overlapping polygons of the buffered nodes.
     """
-    # buffer nodes GeoSeries then get unary union to merge overlaps
-    merged = convert.graph_to_gdfs(G, edges=False)["geometry"].buffer(tolerance).unary_union
+    gdf_nodes = convert.graph_to_gdfs(G, edges=False)
+
+    if isinstance(tolerance, dict):
+        # Map tolerances to node IDs, using NaN for nodes not in the dictionary, and fill NaN with zero
+        buffer_distances = gdf_nodes.index.to_series().map(tolerance).fillna(0)
```

Comment:
Just to ensure you are aware - buffering a point with a distance 0 returns POLYGON EMPTY. After union, some points will not be within any geometry. I am not sure if that is an issue or not, just wanted to flag this. The left sjoin that happens later will result in many points with cluster None. I would just double check that these are not actually considered one cluster and merged together.

This certainly needs a test case with missing keys in the dict.

Reply:
Agreed. I won’t have time in the near future to write those, is anyone wants to continue from this please do so!

Reply:
See my code suggestion at https://github.com/gboeing/osmnx/pull/1160#discussion_r1578804479

It essentially buffers missing nodes from dict with None then fills resulting null geometries with the original point geometries. This was meant to avoid empty geometries. @martinfleis does this address the issue you're raising?

Reply:
In principle yes, that code not really. `fillna` does not see POLYGON EMPTY as missing values, so you need an explicit mask for that.

Reply:
@martinfleis to clarify, it was this snippet:

```python
# ...if the passed tolerance argument is type dict instead of float
gdf_nodes = ox.convert.graph_to_gdfs(G, edges=False)
tolerances = pd.Series(tolerance).reindex(gdf_nodes.index)
gdf_nodes.buffer(tolerances).fillna(gdf_nodes["geometry"])
```

It does not buffer a missing node's point with distance 0 (which would indeed result in POLYGON EMPTY). It instead buffers them with None (which results in None, right?). Then `fillna` should work.

Reply:
Just ran a quick local test and it appears to be so. The result is a collection of buffer Polygons (when node tolerance is missing in dict) and original Points (when it is missing).

Reply:
Ah, yes. I got confused by `fillna(0)` in this PR. Yes, this should resolve it and given the point is within self, the sjoin later should be fine.

Reply:
Implemented!

Thanks for catching this.

Reply:
Thanks @martinfleis. @EwoutH I'll take a final review this morning and provide any final comments.